README.md
TASHY is short from Tcl Ada SHell Younger. It is derivate of [TASH](https://github.com/simonjwright/tcladashell)

**IMPORTANT:** TASHY is not compatible with TASH. Here are many breaking
changes in the code, thus TASHY should be used only with the new code. At
this moment TASHY API is stable. All changes (compared with TASH too) can
be found [in CHANGELOG.md](https://www.laeran.pl/repositories/tashy/file?name=CHANGELOG.md&ci=tip).  If you read this
file on Github: **please don't send pull requests here**. All will be
automatically closed. Any code propositions should go to the [Fossil](https://www.laeran.pl/repositories/tashy)
repository.

TASHY is the Ada 2012 binding to Tcl/Tk. It allows using Tcl code in Ada code
and vice versa. Included bindings:

* Tcl: It is the base binding. Allow use Tcl scripts in the Ada code.
  This binding should be complete but have lack of the code documentation. If
  you find that something is missing, please [open a ticket](https://www.laeran.pl/repositories/tashy/tktnew) with it.
* Tk: Requires Tcl binding. Allows use Tk GUI in the Ada code. At this
  moment it is a low level binding: mostly works like standard Tcl/Tk
  commands. You can use any Tk command/widget by `Tcl_Eval` or
  `Execute_Widget_Command` procedures. If you find that something is
  missing, please [open a ticket](https://www.laeran.pl/repositories/tashy/tktnew) with it.
* Tklib: Requires Tcl and Tk bindings. Allows use additional packages for
  Tk GUI in the Ada code. At this moment it is a low level binding only to
  tooltip, getstring, autoscroll and ctext packages. If you want binding to
  any other package from this library, feel free to [open a ticket](https://www.laeran.pl/repositories/tashy/tktnew) with it.
* Msgcat: Requires Tcl binding. Allows creating multi-language text messages.
  At this moment it is a low level binding to msgcat commands: mostly works
  like standard Tcl commands. If you want binding to any other command from
  this package, please [open a ticket](https://www.laeran.pl/repositories/tashy/tktnew) with it.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

Tag **BREAKING:** means that change breaks compatibility with earlier version
of the project.

## [8.6.12] - 2021-09-08

### Added
- Constant Null_Interp
- Procedure `Dialog` to `Dialogs` package
- Option to build only console version of binding (Tcl + msgcat)

### Changed
- **BREAKING:** Type Tcl_Interp to System.Address. In the most cases it
  should'nt not break anything.
- Updated code documentation
- Ability to use the library in SPARK code
- Default compilation flags for the library

## [8.6.11] - 2020-12-18

### Changed
- Replaced using Widget_Image in code on just Widget in many subprograms
- Updated README.md
- Moved contributing guide to the project documentation
- Moved code of conduct to the project documentation

### Fixed
- Raising Tk\_Widgets commands (Tcl.Tk.Ada.Widgets.Widget\_Raise)
- Lowering Tk\_Widgets below another selected (Tcl.Tk.Ada.Widgets.Lower)
- Typos and grammar in README.md

## [8.6.10] - 2020-10-13

### Changed
- Updated code documentation

## [8.6.9] - 2020-09-30

### Added
- Function `Get_Widget` to get existing Tk widgets

### Changed
- Updated code documentation
- Updated README.md

## [8.6.8] - 2020-08-31

### Added
- Binding for Tk command: `bind tag sequence`

### Changed
- Updated code documentation
- Removed generating debugging information from default compilation flags
- Updated README.md

### Fixed
- Code documentation for Tcl.Tklib.Ada.GetString.Tk\_Get\_String and
  Tcl.Tk.Ada.Widgets.TtkEntry.TtkComboBox and Tcl.Tk.Ada.Widgets.Listbox.Yview
- Setting transparency (Tcl.Tk.Image.Photo.Transparency\_Set) for photo images

## [8.6.7] - 2020-07-31

### Added
- Biding for Tk `clipboard` commands: `append`, `clear`, `get`
- Binding for msgcat commands: `mc`, `mclocale`, `mcload`, `mcset`, `mcmax`,
  `mcexists`, `mcpreferences`, `mcloadedlocales get`, `mcforgetpackage`,
  `mcloadedlocales present`, `mcloadedlocales clear`, `mcmset`
- Binding for Tk `selection` commands: `clear`, `get`, `handle`, `own` (get
  name of owner of selection), `own` (set owner of selection)
- Binding for Tklib `autoscroll` commands: `wrap`, `unwrap`, `autoscroll`,
  `unautoscroll`
- Binding for Tk `option` command: `add`, `clear`, `get`, `readfile`
- Binding for Tklib `getstring` package
- Binding for Tk command: `bell`, `bindtags`, `tk caret`, `tk inactive`,
  `tk scaling`, `tk useinputmethods`
- Binding for Tklib `tooltip` commands: `clear`, `disable`, `enable`, `delay`,
  `fade`
- Binding for Tklib `ctext` commands: `ctext`, `addHighlightClass`,
  `addHighlightClassWithOnlyCharStart`, `addHighlightClassForSpecialChars`,
  `addHighlightClassForRegexp`, `clearHighlightClasses`, `getHighlightClasses`,
  `deleteHighlightClass`, `enableComments`, `disableComments`, `highlight`,
  `fastdelete`, `fastinsert`, `copy`, `cut`, `paste`, `append`

### Changed
- Updated code documentation
- Updated README.md

## [8.6.6] - 2020-06-30

### Added
- Binding for Tk `ttk::notebook` commands: `insert`, `tab`, `tabs`
- Binding for Tk `ttk::panedwindow` commands: `identify element`,
  `identify sash`, `pane`, `sashpos`
- Binding for Tk `ttk::progressbar` commands: `start`, `stop`
- Binding for Tk `ttk::scrollbar` commands: `delta`, `fraction`, `get`, `set`
- Binding for Tk `ttk::sizegrip` widget
- Binding for Tk `ttk::spinbox` commands: `current`, `set`
- Binding for Tk `ttk::treeview` commands: `identify region`,
  `identify column`, `identify element`, `identify item`, `see`, `set`,
  `tag add`,`tag bind`, `tag configure`, `tag has`, `tag names`,
  `tag remove`, `xview`, `xview moveto`, `xview scroll`, `yview`,
  `yview moveto`, `yview scroll`
- Binding for Tk commands: `identify element`, `tk_popup`, `tkwait`,
  `tk_chooseColor`, `tk_fontchooser configure`, `tk_fontchooser show`,
  `tk_fontchooser hide`, `tk_focusNext`, `tk_focusPrev`, `tk_focusFollowsMouse`
- Option to save the project settings in setup script to the file (issue #3)
- Binding for Tk `font` commands: `actual`, `configure`, `create`, `delete`,
  `families`, `measure`, `metrics`, `names`
- Binding for Tk `canvas` command `create` as function
- Binding for Tk `event` commands: `add`, `delete`, `generate`, `info`
- Binding for Tk `grab` commands: `current`, `release`, `set`, `status`

### Changed
- Updated code documentation
- Updated README.md

### Fixed
- Confusion about available bindings for build in setup script (issue #1)
- Don't display unnecessary settings on Windows systems (issue #2)
- The setup script doesn't supports spaces in paths (issue #4)
- The setup script doesn't use entered paths (issue #5)

## [8.6.5] - 2020-05-31

### Added
- Binding for Tk `spinbox` commands: `selection element`, `set`
- Binding for Tk `entry` commands: `bbox`, `icursor`, `index`, `scan mark`,
  `scan dragto`, `seletion adjust`, `selection clear`, `selection from`,
  `selection present`, `selection range`, `selection to`, `validate`, `xview`
  `xview moveto`, `xview scroll`
- Binding for Tk `text` commands: `bbox`, `compare`, `count`, `debug`,
  `dlineinfo`, `dump`, `edit canredo`, `edit canundo`, `edit modified`,
  `edit redo`, `edit reset`, `edit separator`, `edit undo`, `get`,
  `image cget`, `image configure`, `image create`, `image names`, `index`,
  `mark gravity`, `mark names`, `mark next`, `mark previous`, `mark set`,
  `mark unset`, `peer create`, `peer names`, `pendingsync`, `replace`,
  `scan mark`, `scan dragto`, `search`, `see`, `sync`, `tag bind`, `tag cget`,
  `tag delete`, `tag names`, `tag nextrange`, `tag prevrange`, `tag raise`,
  `tag range`, `tag remove`, `window cget`, `window configure`,

src/tclmacro.c
/*
 * tclmacro.c --
 *
 *    This file encapsulates calls to all tcl.h macro functions into C
 *    function calls.  These can then be called from Ada.  This avoids
 *    having to translate the macro.
 *
 *  Copyright (c) 1999-2000 Terry J. Westley
 *
 *  Tash is free software; you can redistribute it and/or modify it
 *  under terms of the GNU General Public License as published by the
 *  Free Software Foundation; either version 2, or (at your option)
 *  any later version. Tash is distributed in the hope that it will be
 *  useful, but WITHOUT ANY WARRANTY; without even the implied
 *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 *  PURPOSE. See the GNU General Public License for more details. You
 *  should have received a copy of the GNU General Public License
 *  distributed with Tash; see file COPYING. If not, write to
 *
 *          Free Software Foundation
 *          59 Temple Place - Suite 330
 *          Boston, MA 02111-1307, USA
 *
 *  As a special exception, if other files instantiate generics from
 *  this unit, or you link this unit with other files to produce an
 *  executable, this unit does not by itself cause the resulting
 *  executable to be covered by the GNU General Public License. This
 *  exception does not however invalidate any other reasons why the
 *  executable file might be covered by the GNU Public License.
 */

#if 0
#define __CYGWIN__
#endif

#include <tcl.h>
#include <stdlib.h>

int Tcl_CallIncrRefCount (struct Tcl_Obj *objPtr)
{
   return Tcl_IncrRefCount (objPtr);
}

void Tcl_CallDecrRefCount (struct Tcl_Obj *objPtr)
{
   Tcl_DecrRefCount (objPtr);
}

int Tcl_CallIsShared (struct Tcl_Obj *objPtr)
{
   return Tcl_IsShared (objPtr);
}

ClientData Tcl_CallGetHashValue (Tcl_HashEntry *h)
{
   return Tcl_GetHashValue (h);
}

void Tcl_CallSetHashValue (Tcl_HashEntry *h, ClientData value)
{
   Tcl_SetHashValue (h, value);
}

char *Tcl_CallGetHashKey (Tcl_HashTable *tablePtr, Tcl_HashEntry *h)
{
   return Tcl_GetHashKey (tablePtr, h);
}

Tcl_HashEntry *Tcl_CallFindHashEntry (Tcl_HashTable *tablePtr, char *key)
{
   return Tcl_FindHashEntry (tablePtr, key);
}

Tcl_HashEntry *Tcl_CallCreateHashEntry
 (Tcl_HashTable *tablePtr, char *key, int *newPtr)
{
   return Tcl_CreateHashEntry (tablePtr, key, newPtr);
}





scripts/generatedocs.py
#!/usr/bin/env python
# Copyright (c) 2019 Bartek thindil Jasicki <thindil@laeran.pl>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
""" Generate documentation and fix syntax highlightning for Ada """

import glob
import re
import subprocess
import sys

def fixdocs(directory):
    """ Fix all generated documentation in selected directory. """
    for filename in glob.glob(directory + "/*.html"):
        # read whole file to list of lines
        with open(filename) as fn:
            content = fn.readlines()
        for i in range(len(content)):
            result = re.search("<span class=\"squote\">'\w+'*", content[i])
            if result is not None:
                keyword = result.group()[21:]
                if keyword[len(keyword) - 1] != "'":
                    content[i] = re.sub("<span class=\"squote\">'\w+", "<span class=\"keyword\">" + keyword + "</span>", content[i]);
                k = i
                while content[k].strip() != "</pre>":
                    result2 = re.search("(?!\">)(\)|;|:=|:|\()+(?!</span>)", content[k])
                    while result2 is not None:
                        endline = content[k][result2.span()[1]:]
                        content[k] = content[k][:result2.span()[0]] + "<span class=\"sign\">"
                        for j in range(len(result2.group())):
                            content[k] += result2.group()[j]
                        content[k] += "</span>" + endline
                        result2 = re.search("(?!\">)(\)|;|:=|:|\()+(?!</span>)", content[k])
                    result2 = re.search("(?!\">)(\sis\s|\spragma\s|\sreturn\s|\sconstant\s)+(?!</span>)", content[k])
                    while result2 is not None:
                        endline = content[k][result2.span()[1]:]
                        content[k] = content[k][:result2.span()[0]] + "<span class=\"keyword\">"
                        for j in range(len(result2.group())):
                            content[k] += result2.group()[j]
                        content[k] += "</span>" + endline
                        result2 = re.search("(?!\">)(\sis\s|\spragma\s|\sreturn\s|\sconstant\s)+(?!</span>)", content[k])
                    k += 1
                    if k == len(content):
                        break
        # save fixed file
        with open(filename, "w") as newfile:
            for line in content:
                newfile.write("%s" % line)
        # fix documentation in any subdirectory too
    for dirname in glob.glob(directory + "*/"):
        fixdocs(dirname)

docsdir = "docs"
configfile = "scripts/robodocada.rc"

if len(sys.argv) > 1:
    configfile = sys.argv[1]
    if len(sys.argv) > 2:
        docsdir = sys.argv[2]

subprocess.call(["robodoc", "--rc", configfile])
fixdocs(docsdir)


scripts/setup.tcl
#!/bin/sh
# the next line restarts using wish \
exec wish "$0" ${1+"$@"}

# Copyright (C) 1997-2000 Terry J. Westley
# Copyright (C) Simon Wright <simon@pushface.org>
# Copyright (C) 2019-2021 Bartek Jasicki <thindil@laeran.pl>

# This package is free software; you can redistribute it and/or
# modify it under terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2, or
# (at your option) any later version. This package is distributed in
# the hope that it will be useful, but WITHOUT ANY WARRANTY; without
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU General Public License for more
# details. You should have received a copy of the GNU General Public
# License distributed with this package; see file COPYING.  If not,
# write to the Free Software Foundation, 59 Temple Place - Suite
# 330, Boston, MA 02111-1307, USA.

# This is a Tcl/Tk script which helps install TASHY.  It collects
# information about the environment and creates file tash_options.gpr,
# which is withed by GPRs to customize to the local environment. Note:
# Please don't add "package require Tk" here. This script is designed
# to work with headless servers also. Manually run it only with "wish"
# command

set buildoption all
set installtklib 0
set tashy_version 8.6.12
set savesettings 0
set installmsgcat 0

# Used for mostly for Docker, normally Tk should set this variable
# by itself
if {[info exists tk_version] == 0} {
   set tk_version 8.6
}

proc cequal {left right} {
   return [expr [string compare $left $right] == 0]
}

proc lempty {string} {
   return [expr [string length $string] == 0]
}

proc setvar {name value comments} {
   global tashvar tashorder tashcomments
   set tashvar($name) $value
   set tashcomments($name) $comments
   lappend tashorder $name
}

# Create the tash_options.gpr file
#--------------------------------
proc CreateGprFile {} {
   global library_switches gpr tashvar installtk
   if [catch {open $gpr w} gprfid] {
      puts stderr $gprfid
      exit
   }
   set library_switch [string range $library_switches 0 [string first "-ltcl" $library_switches]-1]
   puts $gprfid "--  $gpr
   ----------------------------------------------------------
   --
   --  This file, tash_options.gpr, contains global options
   --  used for building and using Tashy.
   --
   --  It is automatically generated by the setup.tcl script.
   --
   ----------------------------------------------------------

   project Tash_Options is

   for Source_Dirs use ();

   -- The source files used to compile Tashy
   Source_Files := ($tashvar(SOURCES));

   --  These are the Ada compiler options used to build Tashy.
   Compiler_Options :=
   (
   \"[join $tashvar(AARGS) "\",\n      \""]\"
   );

   --  These are the C compiler options used to build Tashy.
   C_Compiler_Options :=
   (
   \"[join [concat $tashvar(CARGS)] "\",\n      \""]\",
      \"[join [list "$tashvar(TCL_INCLUDE)" "$tashvar(X11_INCLUDE)"] "\",\n      \""]\"
   );

   --  These options determine the location of the system's Tcl, Tk
   --  libraries.
   Linker_Options :=
   (
   \"[join $tashvar(LARGS) "\",\n      \""]\"
   );

   end Tash_Options;"
   catch {close $gprfid}
}

# Implement Save button command.  Update tashorder & tashvar from the
# GUI.
#-----------------------------------------------------------------
proc Save_GUI {g} {
   global tashorder tashvar useLinkerOptions tcl_platform library_switches
   global tcl_version tk_version
   foreach name $tashorder {
      switch -regexp $name {
         "(TK|TCL)_(VERSION|RELEASE)" {
            # wasn't displayed, can't have changed
         }
         "EXE" {
            # wasn't displayed, can't have changed
         }
         default {
            set w [string tolower $name]
            if {![winfo exists $g.$w-entry]} continue;
            set tashvar($name) "[$g.$w-entry get]"
         }
      }
   }
   set library_switches "-fpic -lto"
   set tclhome $tashvar(TCLHOME)
   regsub -all {[ \t]+} $tcl_platform(os) "_" os
   switch $tcl_platform(platform) {
      "windows" {
         regsub {\.} $tcl_version {} tcl_short_version
         regsub {\.} $tk_version  {} tk_short_version
         append library_switches "-L$tclhome/lib "
         set tail " "
         if {![file exists [file join $tclhome lib tcl$tk_short_version.lib]]} {
            set tail "t "
         }
         append library_switches "-ltcl$tcl_short_version$tail"
         append library_switches "-ltk$tk_short_version$tail"
      }
      "unix" {
         if [cequal $os "SunOS"] {
            append library_switches " -R$tclhome/lib -L$tclhome/lib"
            append library_switches " -ltcl$tcl_version -ltk$tk_version"
         } elseif [cequal $os "Darwin"] {
            append library_switches " -L$tclhome/lib"
            append library_switches " -ltcl$tcl_version -ltk$tk_version"
         } else {
            # Must be Linux
            append library_switches " -ltcl$tcl_version -ltk$tk_version"
         }
      }
   }
   Save
}

# Implement Save command.  Creates tash_options.gpr
# Also, create tcl_record_sizes.ads file
#-----------------------------------------------------------------
proc Save {} {
   global tashvar tcl_platform buildoption library_switches installtklib
   global savesettings installmsgcat argv
   if {$buildoption == "tcl"} {
      setvar TK_VERSION    ""        {Tk version}
      setvar TK_LIBRARY    ""        {Tk library}
      setvar X11HOME       ""        {X11 home directory}
      setvar X11_LIB       ""        {X11 library directory}
      setvar X11_INCLUDE   ""        {X11 include directory}
      if {$installmsgcat == 1} {
         setvar SOURCES "\"src\", \"src/msgcat\"" {Source directories}
      } else {
         setvar SOURCES       "\"src\""   {Source directories}
      }
      set library_switches [string range $library_switches 0 [string first "-ltk" $library_switches]-1]
   } elseif {$buildoption != "all"} {
      set sources "\"src\", \"src/tk\""
      if {$installtklib == 1} {
         append sources ", \"src/tklib\""
      }
      if {$installmsgcat == 1} {
         append sources ", \"src/msgcat\""
      }
      setvar SOURCES       $sources   {Source directories}
   } else {
      setvar SOURCES       "\"src/**\""   {Source directories}
   }
   CreateGprFile
   source [file join [pwd] scripts tcl_record_sizes.tcl]
   set CC gcc
   set argument [lsearch -inline $argv "--CC=*"]
   if {$argument != {}} {
      set CC [string range $argument 5 [string length $argument]]
   }
   exec $CC -Wno-format $tashvar(TCL_INCLUDE) -o tcl_record_sizes [file join [pwd] src tcl_record_sizes.c]
   file delete [file join [pwd] src tcl_record_sizes.c]
   exec [file join [pwd] tcl_record_sizes] > [file join [pwd] src tcl_record_sizes.ads]
   file delete [file join [pwd] tcl_record_sizes$tashvar(EXE)]
   if {$savesettings == 1} {
      set f [open "settings.txt" w]
      puts $f $tashvar(PLATFORM)
      puts $f $tashvar(OS)
      puts $f $tashvar(X11HOME)
      puts $f $tashvar(X11_LIB)
      puts $f $tashvar(X11_INCLUDE)
      puts $f $tashvar(TCLSH)
      puts $f $tashvar(TCLHOME)
      puts $f $tashvar(TCL_INCLUDE)
      puts $f $tashvar(TCL_VERSION)
      puts $f $tashvar(TCL_LIBRARY)
      puts $f $tashvar(TK_VERSION)
      puts $f $tashvar(TK_LIBRARY)
      puts $f $tashvar(CARGS)
      puts $f $tashvar(AARGS)
      puts $f $tashvar(BARGS)
      puts $f $tashvar(EXE)
      puts $f $tashvar(LARGS)
      puts $f "$buildoption"
      puts $f "$installtklib"
      puts $f "$installmsgcat"
      close $f
   }
}


# Establish values for all macros depending on platform
#------------------------------------------------------
proc Set_Macros {platform os} {
   global tcl_version tk_version tcl_interactive tcl_library tk_library env
   global library_switches gpr_switches buildoption installtklib

   if {![file exists settings.txt]} {
      set x11home           ""
      set x11_lib           ""
      set x11_include       ""
      set exec_suffix       ""

      regsub -all {[ \t]+} $os "_" os

      if [cequal $os "Darwin"] {
         set tclhome "/usr"
         set tcl_include "/usr/include"
      } else {
         set tclhome [file dirname [file dirname [info nameofexecutable]]]
         set tcl_include [file join $tclhome include]
         if {![file exists [file join $tcl_include tcl.h]]} {
            set tcl_include [file join $tcl_include [file tail $tcl_library]]
         }
      }
      set library_switches  "-fpic "

      set pwd               [pwd]

      switch $platform {
         "windows" {
            # It assumes a GNAT that recognises tcl86.lib as a
            # candidate for the linker switch -ltcl86.
            #
            # Should works with ActiveState Tcl and Magicsplat Tcl
            # distributions
            #
            # Most development tools get confused by paths with spaces.
            regsub {PROGRAM FILES} $tclhome "PROGRA~1" tclhome
            regsub {\.} $tcl_version {} tcl_short_version
            regsub {\.} $tk_version  {} tk_short_version
            set tclsh "tclsh${tcl_short_version}"
            set libtcl ""
            set tcldll "tcl${tcl_short_version}.dll"
            set libtk ""
            set tkdll  "tk${tk_short_version}.dll"
            append library_switches "-L$tclhome/lib "
            set tail " "
            if {![file exists [file join $tclhome lib tcl$tk_short_version.lib]]} {
               set tail "t "
            }
            append library_switches "-ltcl$tcl_short_version$tail"
            append library_switches "-ltk$tk_short_version$tail"
            set exec_suffix ".exe"
         }
         "unix" {
            set tclsh "tclsh"
            set dynlib [info sharedlibextension]
            set libtcl "$tclhome/lib/libtcl${tcl_version}$dynlib"
            set libtk  "$tclhome/lib/libtk${tk_version}$dynlib"

            set PossibleXHomes \
               [list /usr/openwin /usr/X /usr/X11R6 /usr /opt/X11]
            foreach dir $PossibleXHomes {
               set lib [file join $dir lib]
               foreach file [list "libX11$dynlib" "libX11.a"] {
                  if [file exists [file join $lib $file]] {
                     set x11home $dir
                     set x11_lib [file join $x11home lib]
                     break
                  }
               }
            }
            if [file isdirectory [file join $x11home include]] {
               set x11_include [file join $x11home include]
            } else {
               foreach dir $PossibleXHomes {
                  set include [file join $dir include]
                  if [file isdirectory $include] {
                     set x11_include $include
                     break
                  }
               }
            }
            if [cequal $os "SunOS"] {
               append library_switches " -R$tclhome/lib -L$tclhome/lib"
               append library_switches " -ltcl$tcl_version -ltk$tk_version"
            } elseif [cequal $os "Darwin"] {
               append library_switches " -L$tclhome/lib"
               append library_switches " -ltcl$tcl_version -ltk$tk_version"
            } else {
               # Must be Linux
               append library_switches "-ltcl$tcl_version -ltk$tk_version"
            }
         }
      }
      setvar PLATFORM          $platform            {OS platform}
      setvar OS                $os                  {Operating system}
      if [lempty $x11home] {
         setvar X11HOME       ""                   {X11 home directory}
      } else {
         setvar X11HOME       "$x11home"           {X11 home directory}
      }
      if [lempty $x11_lib] {
         setvar X11_LIB       ""                   {X11 library directory}
      } else {
         setvar X11_LIB       "$x11_lib"           {X11 library directory}
      }
      if [lempty $x11_include] {
         setvar X11_INCLUDE   ""                   {X11 include directory}
      } else {
         setvar X11_INCLUDE   "-I$x11_include"     {X11 include directory}
      }
      setvar TCLSH             "$tclsh"             {Tclsh executable}
      setvar TCLHOME           "$tclhome"           {Tcl Home directory}
      setvar TCL_INCLUDE       "-I$tcl_include"     {Tcl include directory}
      setvar TCL_VERSION       "$tcl_version"       {Tcl version}
      setvar TCL_LIBRARY       "$libtcl"            {Tcl library}
      setvar TK_VERSION        "$tk_version"        {Tk version}
      setvar TK_LIBRARY        "$libtk"             {Tk library}
      setvar CARGS             "-O2 -fpic -lto"          {C compiler switches}
      setvar AARGS             "-O2 -gnatafo -gnatwaL -fpic -lto" \
         {Ada compiler switches}
      setvar BARGS             "-E"                 {gnatbind switches}
      setvar EXE               "$exec_suffix"       {suffix for executable files}
      setvar LARGS             "$library_switches"  {linker switches}
   } else {
      set f [open "settings.txt"]
      setvar PLATFORM          "[gets $f]"   {OS platform}
      setvar OS                "[gets $f]"   {Operating system}
      setvar X11HOME           "[gets $f]"   {X11 home directory}
      setvar X11_LIB           "[gets $f]"   {X11 library directory}
      setvar X11_INCLUDE       "[gets $f]"   {X11 include directory}
      setvar TCLSH             "[gets $f]"   {Tclsh executable}
      setvar TCLHOME           "[gets $f]"   {Tcl Home directory}
      setvar TCL_INCLUDE       "[gets $f]"   {Tcl include directory}
      setvar TCL_VERSION       "[gets $f]"   {Tcl version}
      setvar TCL_LIBRARY       "[gets $f]"   {Tcl library}
      setvar TK_VERSION        "[gets $f]"   {Tk version}
      setvar TK_LIBRARY        "[gets $f]"   {Tk library}
      setvar CARGS             "[gets $f]"   {C compiler switches}
      setvar AARGS             "[gets $f]"   {Ada compiler switches}
      setvar BARGS             "[gets $f]"   {gnatbind switches}
      setvar EXE               "[gets $f]"   {suffix for executable files}
      setvar LARGS             "[gets $f]"   {linker switches}
      set buildoption "[gets $f]"
      set installtklib "[gets $f]"
      set installmsgcat "[gets $f]"
      close $f
   }
}

# Establish values for all macros depending on platform
#------------------------------------------------------
Set_Macros $tcl_platform(platform) $tcl_platform(os)

set gpr [file join [pwd] tash_options.gpr]

# If no GUI required, save and finish
#------------------------------------
if {[lindex $argv 0] == "--nogui"} {
   Save
   exit
}
# Build only Tcl and its children (no GUI)
#-----------------------------------------
if {[lindex $argv 0] == "--headless"} {
   set buildoption "tcl"
   set installmsgcat "1"
   Save
   exit
}

# Create window for installer to review and edit macro values
#------------------------------------------------------------
wm title    . "TASHY $tashy_version Setup"
wm iconname . "TASHY Setup"

message .instructions -justify left -aspect 500 -pady 10 -padx 20 -text \
   "This program, setup.tcl, customizes the TASHY $tashy_version installation by\
   creating the file $gpr. Setup.tcl guesses \"reasonable\"\
   values for the macros, but you may have to edit them.  After you're\
   happy with the macro values, press \"Save\" to save the files."

pack .instructions -side top -fill x

set modulesframe [ttk::labelframe .modules -text "Select modules"]
pack $modulesframe -side top

pack [ttk::radiobutton .modules.all -text "Everything" -value "all" -variable buildoption] -fill x
pack [ttk::radiobutton .modules.tcl -text "Tcl and one from the list below" -value "tcl" -variable buildoption] -fill x
pack [ttk::checkbutton .modules.msgcat -text "msgcat" -variable installmsgcat]
pack [ttk::radiobutton .modules.select -text "Tcl, Tk and one from the list below" -value "tcltk" -variable buildoption]
pack [ttk::checkbutton .modules.tklib -text "Tklib" -variable installtklib]
pack [ttk::checkbutton .modules.msgcat2 -text "msgcat" -variable installmsgcat]

set g [ttk::frame .grid]
pack $g -side top

set row 0

proc DisplaySetting {name} {
   global row g w tashvar

   set w [string tolower $name]
   ttk::label $g.$w-label -text "$name: " -anchor e
   grid $g.$w-label  -row $row -column 0 -sticky e
   ttk::entry $g.$w-entry -width 40
   $g.$w-entry insert end $tashvar($name)
   grid $g.$w-entry  -row $row -column 1
   incr row
}

foreach name $tashorder {
   switch -regexp $name {
      "(TK|TCL)_(VERSION|RELEASE)" {
         # don't display
      }
      "EXE" {
         # don't display
      }
      "(X11)(HOME|_LIB|_INCLUDE)" {
         # don't display on Windows
         if {$tcl_platform(platform) != "windows"} {
            DisplaySetting $name
         }
      }
      "(TK|TCL)_(LIBRARY)" {
         # don't display on Windows
         if {$tcl_platform(platform) != "windows"} {
            DisplaySetting $name
         }
      }
      default {
         DisplaySetting $name
      }
   }
}

pack [ttk::checkbutton .savesetting -text {Save these settings to the file} -variable savesettings]
ttk::frame .buttons
pack .buttons -side bottom -fill x -pady 2m
ttk::button .buttons.save   -text Save   -command "Save_GUI $g; exit"
ttk::button .buttons.cancel -text Cancel -command exit
pack .buttons.save .buttons.cancel -side left -expand 1
